docs/12.EXT-REFERENCE-MODULE.md

```diff
@@ -0,0 +1,182 @@
+#  Module API
+
+**Note**: This extension depends on compiler support because it uses `__attribute__` extensions provided only by GCC and
+LLVM/clang. For other compilers, no support has been added yet.
+
+A JerryScript module is defined as a global static structure (`jerryx_module_t`) consisting of a string representing the
+name of the module and a function pointer that returns a `jerry_value_t` representing an instance of the module.
+
+A module resolver (`jerryx_module_resolver_t`) is a function which, given a string representing a module name, will fill
+out its `jerry_value_t *` out-parameter with the result of having loaded the module bearing the requested name after
+having found it.
+
+Since different types of modules may be structured differently, for each type of module a module resolvers must be
+supplied at the point where an instance of that type of module is requested. `jerryx_module_native_resolver ()` is a
+resolver for JerryScript modules as defined above.
+
+## jerryx_module_resolve
+
+**Summary**
+
+Load a copy of a module into the current context or return one that was already loaded if it is found.
+
+Each function in `resolvers_p` will be called in sequence until one returns `true` and fills out its out-parameter with
+the `jerry_value_t` representing the requested module. If the `jerry_value_t` does not have the error flag set it will
+be cached. Thus, on subsequent calls with the same value for `name`, none of the functions in `resolvers_p` will be
+called.
+
+**Prototype**
+
+```c
+jerry_value_t
+jerryx_module_resolve (const jerry_char_t *name,
+                       *resolvers_p,
+                       size_t resolver_count);
+```
+
+- `name` - the name of the module to load
+- `resolvers_p` - the list of resolvers to call in sequence
+- `resolver_count` - the number of resolvers in `resolvers_p`
+- return value - `jerry_value_t` representing the module that was loaded, or the error that occurred in the process.
+
+
+## jerryx_module_native_resolver
+
+**Summary**
+
+The resolver for JerryScript modules. A pointer to this function can be passed in the second parameter to
+`jerryx_module_resolve` to search for the module among the JerryScript modules built into the binary.
+
+**Prototype**
+
+```c
+bool jerryx_module_native_resolver (const jerry_char_t *name,
+                                    jerry_value_t *result)
+```
+- `name` - the name of the module to find
+- `result` - out - place where to store the resulting module instance
+- return value - `true` if the module was found and stored in `result`, and `false` otherwise
+
+
+# Module data types
+
+## jerryx_module_t
+
+**Summary**
+
+Structure to define a JerryScript module
+
+**Prototype**
+
+```c
+typedef struct
+{
+  jerry_char_t *name; /**< name of the module */
+  jerry_value_t (*on_resolve) (void); /**< function that returns a new instance of the module */
+} jerryx_module_t;
```

Comment:
I think this structure is only used by native modules. Furthermore do we need to make it public? I mean it is needed to be public, but people should never use its members directly. We should emphasize this.